File: condition-api/src/condition_api/schemas/condition.py
"""Condition model class.

Manages the condition
"""

from condition_api.schemas.condition_attribute import ConditionAttributesSchema
from condition_api.schemas.subcondition import SubconditionSchema

from marshmallow import Schema, fields


class ConditionSchema(Schema):
    """Condition schema."""

    condition_id = fields.Str(data_key="condition_id")
    condition_name = fields.Str(data_key="condition_name", allow_none=True)
    condition_number = fields.Int(data_key="condition_number", allow_none=True)
    condition_text = fields.Str(data_key="condition_text", allow_none=True)
    topic_tags = fields.List(fields.Str(), data_key="topic_tags", allow_none=True)
    subtopic_tags = fields.List(fields.Str(), data_key="subtopic_tags", allow_none=True)
    amendment_names = fields.Str(data_key="amendment_names", allow_none=True)
    year_issued = fields.Int(data_key="year_issued", allow_none=True)
    is_approved = fields.Bool(data_key="is_approved", allow_none=True)
    is_topic_tags_approved = fields.Bool(data_key="is_topic_tags_approved", allow_none=True)
    is_condition_attributes_approved = fields.Bool(
        data_key="is_condition_attributes_approved", allow_none=True)
    requires_management_plan = fields.Bool(
        data_key="requires_management_plan", allow_none=True)
    condition_attributes = fields.Nested(ConditionAttributesSchema, data_key="condition_attributes")
    effective_document_id = fields.Str(data_key="effective_document_id", allow_none=True)
    is_standard_condition = fields.Bool(data_key="is_standard_condition", allow_none=True)
    source_document = fields.Str(data_key="source_document", allow_none=True)
    # Condition can also have its own subconditions (recursive nesting)
    subconditions = fields.List(fields.Nested(SubconditionSchema), data_key="subconditions")


class ProjectDocumentConditionSchema(Schema):
    """Top-level schema to include project and document names."""

    project_name = fields.Str(data_key="project_name")
    document_category = fields.Str(data_key="document_category")
    conditions = fields.List(fields.Nested(ConditionSchema), data_key="conditions")


class ProjectDocumentConditionDetailSchema(Schema):
    """Top-level schema to include project and document names."""

    project_id = fields.Str(data_key="project_id")
    project_name = fields.Str(data_key="project_name")
    document_category_id = fields.Str(data_key="document_category_id")
    document_category = fields.Str(data_key="document_category")
    document_label = fields.Str(data_key="document_label")
    document_id = fields.Str(data_key="document_id")
    condition = fields.Nested(ConditionSchema, data_key="condition")


class ConsolidatedConditionSchema(Schema):
    """Condition schema."""

    condition_name = fields.Str(data_key="condition_name", allow_none=True)
    condition_number = fields.Int(data_key="condition_number", allow_none=True)
    condition_text = fields.Str(data_key="condition_text", allow_none=True)
    topic_tags = fields.List(fields.Str(), data_key="topic_tags", allow_none=True)
    subtopic_tags = fields.List(fields.Str(), data_key="subtopic_tags", allow_none=True)
    amendment_names = fields.Str(data_key="amendment_names", allow_none=True)
    year_issued = fields.Int(data_key="year_issued", allow_none=True)
    condition_attributes = fields.Dict(data_key="condition_attributes", allow_none=True)

    # Condition can also have its own subconditions (recursive nesting)
    subconditions = fields.List(fields.Nested(SubconditionSchema), data_key="subconditions")

File: condition-web/src/models/Condition.ts
import { SubconditionModel } from "./Subcondition";
import { ConditionAttributeModel } from "./ConditionAttribute";

export interface ConditionModel {
  condition_id?: number;
  condition_name?: string;
  condition_number?: number;
  condition_text?: string;
  amendment_names?: string;
  effective_document_id?: string;
  year_issued?: number;
  is_approved?: boolean;
  topic_tags?: string[];
  is_topic_tags_approved?: boolean;
  is_condition_attributes_approved?: boolean;
  is_standard_condition?: boolean;
  requires_management_plan?: boolean;
  source_document?: string;
  subtopic_tags?: string[];
  subconditions?: SubconditionModel[];   // Nested subconditions
  condition_attributes?: ConditionAttributeModel;
}

export const createDefaultCondition = (): ConditionModel => {
  return {
      condition_id: 0,
      condition_name: '',
      condition_number: 0,
      condition_text: '',
      amendment_names: '',
      effective_document_id: '',
      year_issued: 0,
      is_approved: false,
      topic_tags: [],
      is_topic_tags_approved: false,
      is_condition_attributes_approved: false,
      requires_management_plan: false,
      subtopic_tags: [],
      subconditions: [],
      condition_attributes: {
        independent_attributes: [],
        management_plans: [],
      },
  };
};

export interface updateTopicTagsModel {
  condition_number?: number;
  condition_name?: string;
  topic_tags?: string[];
  is_approved?: boolean;
  is_topic_tags_approved?: boolean;
  is_condition_attributes_approved?: boolean;
  subconditions?: SubconditionModel[],
}

export type PartialUpdateTopicTagsModel = Partial<updateTopicTagsModel>;

export interface ProjectDocumentConditionModel {
  project_name: string;
  document_category: string;
  document_category_id: number;
  document_label: string;
  conditions?: ConditionModel[];
}

export interface ProjectDocumentConditionDetailModel {
  project_id: string;
  project_name: string;
  document_category: string;
  document_category_id: number;
  document_id: string;
  document_label: string;
  condition: ConditionModel;
}

export type ConditionStatus = "true" | "false";
export const CONDITION_STATUS: Record<
  ConditionStatus,
  { value: ConditionStatus; label: string }
> = {
  true: {
    value: "true",
    label: "Approved",
  },
  false: {
    value: "false",
    label: "Awaiting Approval",
  }
};
